source/common/upstream/load_balancer_impl.cc

```diff
@@ -739,12 +742,10 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
       // notification, this will only be stale until this host is next picked,
       // at which point it is reinserted into the EdfScheduler with its new
       // weight in chooseHost().
-      scheduler.edf_->add(hostWeight(*host), host);
+      scheduler.edf_->initial_add(hostWeight(*host), host);
```

Comment:
Why is the initial add occurring here and not below? Would the "random pick" behavior you mention in the description occur by cycling through the queue in that loop?

Reply:
It would be too long, so I don't know where to comment.
I'll try to explain here first and then reword my comments.
---
The most simple and ideal way for avoiding bias is random loop pick.
```
if (!hosts.empty()) {
      for (uint32_t i = 0; i < seed_ % SUM_OF_ALL_WEIGHT; ++i) {
        auto host = scheduler.edf_->pick();
```
But random modulo is not `hosts.size()` but `sum_of_all_weights`, which means if we have 1000 endpoints, and every points' weight is close to 100, we have to randomly pick ~[0,100000) times before serving. That is not acceptable for a O(M*N) complexity. (M is num of hosts and N is average weight)
So, my algorithm is try to simulate the heavy random picking scenario.
What would be like when randomly picking 50000 times? Every host's deadline is close to but larger than 0.5 !
So this is the usage for floor().
I pick a random float between 0 and 1, and makes initial deadline close but larger than the random value. The complexity of this algorithm is O(M), much better than O(M*N).